### PR TITLE
FINALLLLYYYY COLOOOOR!

### LINKED ISSUE #540: Allow users to turn off color in the CLI
Add a CLI option to turn color display on in the console. Default value is off for compatibility.

*Acceptance Criteria:*
 * User can use --color to turn color ON in an active TTY session

 * User can use --no-color to override --color set in a config file

### LINKED ISSUE #543: Add formatting and color to logging in CLI
Add the color and formatting to logging in the CLI that is outlined in the CLI Style Guide. This includes color, formatting and organization of content.

*From the style guide:*

General Format:
 * <timestamp> | <log level> | <thread> | <class> | <message>

Color & Text Formatting:
 * Timestamp: Dimmed

 * Log Level - INFO: Cyan - ANSI 96

 * Log Level - ERROR: Bold, Red - ANSI 31

 * Log Level - WARNING: Bold, Bright Yellow - ANSI 33

 * Info Message: No formatting

 * Highlighted Information: Green - ANSI 32

 * Error Message: Bold, Red - ANSI 31

 * Warning Message: Bold, Bright Yellow - ANSI 33

 * Failure Message: Bold, Red - ANSI 31

 * Failure Reasons: Dimmed

 * Thread: Cyan - ANSI 36

 * Class: Cyan - ANSI 36

 * Vertical Bar Separator: Dimmed

Color and formatting is used to help users to easily scan and find relevant information. Cyan is our default color making it easy for users to ignore. This makes red and yellow stand out. Green is used selectively to help call attention to specific information or values in messages.

Spacing & Alignment:
 * Each piece of information is separated by a vertical bar.

 * We keep consistent visual spacing between the log level and the thread/classes (7 spaces) so messages that relate to the same thread and class are visually grouped.

 * INFO is followed by 3 spaces

 * ERROR is followed by 2 spaces

 * WARNING is followed by a single space

 * Spacing and alignment is the same in the command-line and logs to ensure users have a consistent experience.

*Acceptance Criteria:*
 * Color and formatting is only outputted in TTY sessions

 * Formatting matches what is shown in the CLI style guide

### PR DESCRIPTION
**Pull request by RatanRSur**
### PR description
Wow, log4j is way more frustrating than I imagined. But I got it working!
Try it out with `besu --logging=trace`

I need a couple of things from a few people:
* @CjHare According to the git blame for the teku logging, you're the pro on the log4j programmatic configuration and I'd love if you would be the final reviewer ;) If you don't feel like reviewing it feel free to assign someone else
* @atoulme Can you make sure splunk still works? I removed a line from your configuration that could have broken it.
* @benjamincburns Could you make sure this isn't absolutely broken for you on windows?
* **everyone else who wants to bike-shed on this** I only ask that you bike shed by providing an alternate format string which can be easily specified using the format found [here](https://logging.apache.org/log4j/2.x/manual/layouts.html#PatternLayout), scroll down a bit to `highlight`

### Fixed Issue(s)
fixes #543
fixes #540

### CODE REVIEW

**Comment by atoulme** [comment]
If you look in the xml, there are a number of settings set via environment variables at the top, that could be useful to you.
I would recommend that an env variable be set to disable colors if besu identifies it's not running in a tty environment.

**Comment by RatanRSur** [comment]
I think we automatically get env vars for command line options so that still leaves us with the issue of how to change the color behavior at run time.

**Comment by CjHare** [comment]
> issue of how to change the color behavior at run time.
Ah, an opportunity for new JSON-RPC endpoint perhaps 😉

**Comment by CjHare** [comment]
>Ah, an opportunity for new JSON-RPC endpoint perhaps 😉

Sorry, that was a hint/suggestion to use an endpoint to trigger the change in the color printing (similar to that in Teku), rather then leveraging the log4j config file scanning/hot replace feature

**Comment by RatanRSur** [comment]
Looking into how to shift code dependencies around to get around the broken unit test. I have a strong suspicion there's nothing wrong in the production code but there's something about that static variable that the tests don't like 🤔

**Comment by RatanRSur** [comment]
Could you post the contents of the xml file you used? I tried using the two xmls in the teku tree but they didn't cause the double logging.

**Comment by ajsutton** [comment]
I'd have expected anything that defines an appender to have issues.  Mine was:
```
<?xml version="1.0" encoding="UTF-8"?>
<Configuration status="INFO" monitorInterval="5">
    <Appenders>
        <Console name="console" target="SYSTEM_OUT">
            <PatternLayout
                pattern="%white{%d{yyyy-MM-dd HH:mm:ss.SSS}} | %-25.25t | %-5level | %-30.30c{1} | %msg%n"/>
        </Console>
    </Appenders>
    <Loggers>
        <Logger name="teku-event-log" level="INFO"/>
        <!--        <Logger name="tech.pegasys.teku.datastructures.util.GenesisGenerator" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.services.powchain" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.datastructures.util.BeaconStateUtil" level="DEBUG"/>-->
        <!--        <Logger name="tech.pegasys.teku.pow" level="TRACE"/>-->
        <!--        <Logger name="org.ethereum.beacon.discovery" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.datastructures.util.GenesisGenerator" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.statetransition.genesis.GenesisHandler" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.statetransition.attestation" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.validator.coordinator.ValidatorApiHandler" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.validator.client" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.sync" level="INFO"/>-->
        <Logger name="tech.pegasys.teku.statetransition.genesis.GenesisHandler" level="TRACE"/>
        <Logger name="tech.pegasys.teku.pow.MinimumGenesisTimeBlockFinder" level="TRACE"/>
        <Logger name="org.quartz" level="INFO"/>
        <Logger name="io.netty" level="INFO"/>
        <Logger name="io.libp2p" level="INFO"/>
        <!--        <Logger name="tech.pegasys.teku.pow.Eth1HeadTracker" level="DEBUG"/>-->
        <!--        <Logger name="tech.pegasys.teku.networking.eth2.peers.Eth2Peer" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.networking.eth2.peers.PeerChainValidator" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.pow.Web3jEth1Provider" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.networking.p2p.connection.ConnectionManager" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.networking.p2p.libp2p.LibP2PPeer" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.networking.eth2.peers.Eth2PeerManager" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.networking.eth2.rpc.beaconchain.methods.StatusMessageHandler" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler" level="TRACE"/>-->
        <!--        <Logger name="tech.pegasys.teku.networking.eth2.rpc.core.Eth2IncomingRequestHandler" level="DEBUG"/>-->
        <!--        <Logger name="tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler" level="TRACE"/>-->
        <Root level="INFO">
            <AppenderRef ref="console"/>
        </Root>
    </Loggers>
</Configuration>
```

**Comment by RatanRSur** [comment]
Yup, that config caused the double-logging, pushed up changes to address it.

**Comment by RatanRSur** [comment]
@atoulme Did you get a chance to see if the splunk stuff still works?

**Comment by atoulme** [comment]
I didn't. What do you need me to do?

**Comment by RatanRSur** [comment]
If you take a look at the diff, I removed the reference to the appender that you added because it no longer exists at config parse time. It looks suspiciously like it would ruin what you did in your PR

**Comment by RatanRSur** [comment]
@benjamincburns did you get a chance to see if this works on windows?

**Comment by benjamincburns** [comment]
Tested on windows under both powershell and the old dos prompt. I see bash color escape codes on both, so the logs are unreadable there. Following the user-driven workarounds mentioned [here](https://stackoverflow.com/a/51681675/203705) does make colorization work on Windows, though. That said, this isn't set by default, so if this is merged as-is the default UX for windows will suffer some.

**Comment by RatanRSur** [comment]
@ajsutton Any other changes you'd like to see?

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 1]
```diff
@@ -720,6 +720,13 @@ void setBannedNodeIds(final List<String> values) {
       description = "Logging verbosity levels: OFF, FATAL, ERROR, WARN, INFO, DEBUG, TRACE, ALL")
   private final Level logLevel = null;
 
+  @SuppressWarnings({"FieldCanBeFinal"})
+  @Option(
+      names = {"--force-color-enabled"},
```
Comment (ajsutton) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:725]: If there's been discussion about this elsewhere I'm happy to be overruled but personally I would have just called this `--color-enabled`.

**Pull request merged**